File: php/src/vaas/Authentication/ClientCredentialsGrantAuthenticator.php
<?php

namespace VaasSdk\Authentication;

use Amp\Http\Client\HttpClient;

class ClientCredentialsGrantAuthenticator extends TokenReceiver implements AuthenticatorInterface
{
    private readonly string $clientId;
    private readonly string $clientSecret;
    
    /**
     * The authenticator for the client credentials grant type if you have a client id and client secret.
     * @param string $clientId The client id
     * @param string $clientSecret The client secret
     * @param string|null $tokenUrl The optional token url. Defaults to 'https://account.gdata.de/realms/vaas-production/protocol/openid-connect/token'
     * @param HttpClient|null $httpClient Your optional custom http client.
     */
    public function __construct(string $clientId, string $clientSecret, ?string $tokenUrl = null, ?HttpClient $httpClient = null)
    {
        parent::__construct($tokenUrl, $httpClient);
        $this->clientId = $clientId;
        $this->clientSecret = $clientSecret;
    }

    protected function tokenRequestToForm(): string
    {
        return http_build_query([
            'client_id' => $this->clientId,
            'client_secret' => $this->clientSecret,
            'grant_type' => 'client_credentials',
        ]);
    }
}

File: java/src/main/java/de/gdata/vaas/authentication/ClientCredentialsGrantAuthenticator.java
package de.gdata.vaas.authentication;

import lombok.Getter;
import org.jetbrains.annotations.NotNull;

import java.net.URI;
import java.net.URISyntaxException;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.nio.charset.StandardCharsets;

@Getter
public class ClientCredentialsGrantAuthenticator extends TokenReceiver implements IAuthenticator {

    private final String clientId;
    private final String clientSecret;
    private final URI tokenUrl;
    private final HttpClient httpClient;

    /**
     * The authenticator for the client credentials grant type if you have a client id and client secret.
     *
     * @param clientId     The client id
     * @param clientSecret The client secret
     * @param tokenUrl     The optional token url. Defaults to the G DATA production URL
     * @param httpClient   Your optional custom http client.
     */
    public ClientCredentialsGrantAuthenticator(@NotNull String clientId, @NotNull String clientSecret, @NotNull URI tokenUrl, @NotNull HttpClient httpClient) {
        super(tokenUrl, httpClient);
        this.clientId = clientId;
        this.clientSecret = clientSecret;
        this.tokenUrl = tokenUrl;
        this.httpClient = httpClient;
    }

    /**
     * The authenticator for the client credentials grant type if you have a client id and client secret.
     *
     * @param clientId     The client id
     * @param clientSecret The client secret
     * @param tokenUrl     The optional token url. Defaults to the G DATA production URL
     */
    public ClientCredentialsGrantAuthenticator(@NotNull String clientId, @NotNull String clientSecret, @NotNull URI tokenUrl) {
        this(clientId, clientSecret, tokenUrl, HttpClient.newHttpClient());
    }

    /**
     * The authenticator for the client credentials grant type if you have a client id and client secret.
     *
     * @param clientId     The client id
     * @param clientSecret The client secret
     */
    public ClientCredentialsGrantAuthenticator(@NotNull String clientId, @NotNull String clientSecret) throws URISyntaxException {
        this(clientId, clientSecret, new URI("https://account.gdata.de/realms/vaas-production/protocol/openid-connect/token"));
    }

    @Override
    protected String tokenRequestToForm() {
        return "client_id=" + URLEncoder.encode(this.getClientId(), StandardCharsets.UTF_8) +
                "&client_secret=" + URLEncoder.encode(this.getClientSecret(), StandardCharsets.UTF_8) +
                "&grant_type=client_credentials";
    }
}